File: services/product-service/app/src/main/java/io/micro_spring_boot/product_service/app/service/impl/ProductServiceImpl.java
package io.micro_spring_boot.product_service.app.service.impl;

import io.micro_spring_boot.product_service.app.dto.CreateProductRequest;
import io.micro_spring_boot.product_service.app.dto.GetProductRequest;
import io.micro_spring_boot.product_service.app.dto.GetProductResponse;
import io.micro_spring_boot.product_service.app.service.ProductService;
import io.micro_spring_boot.product_service.domain.aggregates.product.Product;
import io.micro_spring_boot.product_service.infra.repositories.ProductRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Slf4j
public class ProductServiceImpl implements ProductService {
    private final ProductRepository productRepository;

    @Override
    public void createProduct(CreateProductRequest productRequest) {
        Product product = new Product(
            productRequest.skuCode(),
            productRequest.name(),
            productRequest.description(),
            productRequest.price(),
            productRequest.type()
        );

        productRepository.save(product);
    }

    @Override
    public List<GetProductResponse> getAllProducts(GetProductRequest request) {
        List<Product> products = request.skuCode() == null ? productRepository.findAll() : productRepository.findBySkuCodeIn(request.skuCode());
        return products.stream().map(this::mapToProductResponse).toList();
    }

    @Override
    public GetProductResponse getProductById(UUID id) {
        Product product = productRepository.findById(id).orElseThrow();
        return mapToProductResponse(product);
    }

    private GetProductResponse mapToProductResponse(Product product) {
        return new GetProductResponse(
            product.getId(),
            product.getSkuCode(),
            product.getName(),
            product.getDescription(),
            product.getPrice(),
            product.getType()
        );
    }
}

File: services/product-service/app/src/main/kotlin/com/microspringboot/productservice/app/services/impl/ProductServiceImpl.kt
package com.microspringboot.productservice.app.services.impl

import com.microspringboot.productservice.app.dtos.CreateProductRequest
import com.microspringboot.productservice.app.dtos.GetProductRequest
import com.microspringboot.productservice.app.dtos.GetProductResponse
import com.microspringboot.productservice.app.services.ProductService
import com.microspringboot.productservice.domain.aggregates.products.Product
import com.microspringboot.productservice.infra.repositories.ProductRepository
import org.springframework.stereotype.Service
import java.util.*

@Service
class ProductServiceImpl(
    private val productRepository: ProductRepository
) : ProductService {
    override fun createProduct(productRequest: CreateProductRequest) {
        val product = Product(
//            UUID.randomUUID(),
            productRequest.skuCode,
            productRequest.name,
            productRequest.description,
            productRequest.price,
            productRequest.type
        )
        productRepository.save(product)
    }

    override fun getAllProducts(request: GetProductRequest): List<GetProductResponse> {
        val products =
            if (request.skuCode == null)
                productRepository.findAll()
            else productRepository.findBySkuCodeIn(request.skuCode)

        return products.stream().map { mapToProductResponse(it) }
            .toList()
    }

    override fun getProductById(id: UUID): GetProductResponse {
        val product = productRepository.findById(id).orElseThrow()
        return mapToProductResponse(product)
    }

    private fun mapToProductResponse(product: Product): GetProductResponse {
        return GetProductResponse(
            product.id,
            product.skuCode,
            product.name,
            product.description,
            product.price,
            product.type
        )
    }
}
